FPGA: Add sample for streaming data interfaces with pipes (#8)
williamwang256/oneAPI-samples: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/streaming_data_interfaces/README.md

# `Streaming Data Interfaces` Sample

> This FPGA sample is a tutorial that demonstrates how to implement streaming data interfaces on an IP component. If you have not already gone over the IP Authoring Interfaces Overview Tutorial, it is recommended that you do so before continuing with this tutorial.

Reviewer: ```suggestion
This FPGA sample is a tutorial that demonstrates how to implement streaming data interfaces on an IP component. It is recommended that you review the IP Authoring Interfaces Overview Tutorial before continuing with this tutorial.
```

| Area                  | Description
|:--                    |:--
| What you will learn   | How to use pipes to implement streaming data interfaces on an IP component
| Time to complete      | 30 minutes
| Category              | Concepts and Functionality

## Purpose

> Pipes are a first-in first-out (FIFO) buffer construct that provide links between elements of a design. They are accessed through read and write APIs without the notion of a memory address or pointer to elements within the FIFO.

Reviewer: ```suggestion
Pipes are a first-in first-out (FIFO) buffer construct that provide data links between elements of a design. They are accessed through read and write APIs without the notion of a memory address or pointers to elements within the FIFO.
```

> The concept of a pipe provides us with a mechanism for specifying streaming data interfaces on an IP component. This tutorial will demonstrate how to declare and configure a pipe to implement your desired interface.

Reviewer: ```suggestion
The concept of a pipe is an intuitive mechanism for specifying streaming data interfaces on an IP component. This tutorial demonstrates how to declare and configure a pipe to configure an Avalon streaming interface.
```

## Prerequisites

| Optimized for        | Description
|:---                  |:---
| OS                   | Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10

> | Hardware             | Intel® Agilex® 7, Arria® 10, and Stratix® 10 FPGAs

Reviewer: ```suggestion
| Hardware             | Intel® Agilex® 7, Cyclone 10® GX, Cyclone V®, Arria® 10, and Stratix® 10 FPGAs
```

   tier1 --> tier2 --> tier3 --> tier4

   style tier1 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier2 fill:#f96,stroke:#333,stroke-width:1px,color:#fff
   style tier3 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier4 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
```

Find more information about how to navigate this part of the code samples in the [FPGA top-level README.md](/DirectProgramming/C++SYCL_FPGA/README.md).
You can also find more information about [troubleshooting build errors](/DirectProgramming/C++SYCL_FPGA/README.md#troubleshooting), [running the sample on the Intel® DevCloud](/DirectProgramming/C++SYCL_FPGA/README.md#build-and-run-the-samples-on-intel-devcloud-optional), [using Visual Studio Code with the code samples](/DirectProgramming/C++SYCL_FPGA/README.md#use-visual-studio-code-vs-code-optional), [links to selected documentation](/DirectProgramming/C++SYCL_FPGA/README.md#documentation), and more.

## Key Implementation Details

### Configuring a Pipe to Implement a Streaming Data Interface

Each individual pipe is a function scope class declaration of the templated `pipe` class.

> - The first template parameter is a user-defined type that differentiates this particular pipe from the others and provides a name for the interface in the  RTL.

Reviewer: This would probably be clearer in a table:

| Template Parameter | Description
|---                 |---
| `name`             | User-defined type that differentiates this particular pipe from the others and provides a name for the interface in the RTL
| `dataT`            | Defines the datatype of elements carried by the interface on the `data` signal *
| `min_capacity`     | Specify a non-negative integer representing the capacity of the buffer on the input. This can help avoid some amount of bubbles in the pipeline in case the component itself stalls. This parameter is optional, and defaults to `0`.
| `properties`       | List of SYCL* `properties` that define additional semantic properties for a pipe. This parameter is optional.**

⚠️ * There is currently a known issue with using struct types whose first field is 8-bits wide, and hence also data types which are themselves 8-bits wide (e.g., unsigned char). For the time being, please use a wider datatype where applicable (e.g., 16-bits).

**A list of those properties relevant to this sample is given in Table 1 (please note that this table is not complete; see the [FPGA Optimization Guide for Intel® oneAPI Toolkits](https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/current/host-pipe-declaration.html) for more information on how to use pipes in other applications).

- The second template parameter defines the datatype of elements carried by the interface.

> > :warning: There is currently a known issue with using struct types whose first field is 8-bits wide, and hence also data types which are themselves 8-bits wide (e.g., `unsigned char`). For the time being, please use a wider datatype where applicable (e.g., 16-bits).

Reviewer 2: I'm guessing that there is a case for this. Please add a note to the case that when addressed, this warning should be removed.

Reviewer: case is here: https://hsdes.intel.com/appstore/article/#/14019166650

I added a reminder about this code sample.

- The third template parameter allows you to optionally specify a non-negative integer representing the capacity of the buffer on the input. This can help avoid some amount of bubbles in the pipeline in case the component itself stalls.
- The fourth template parameter uses the oneAPI properties class to allow users to optionally define additional semantic properties for a pipe. **The `protocol` property is what will allow us to configure a pipe to implement a streaming data interface.** A list of those properties relevant to this sample is given in Table 1 (please note that this table is *not* complete; see the [FPGA Optimization Guide for Intel® oneAPI Toolkits](https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/current/host-pipe-declaration.html) for more information on how to use pipes in other applications).

#### Table 1. Properties used to Configure a Pipe to Implement a Streaming Data Interface

> | Property | Valid Values | Default Value |

Reviewer 2: Please align the table so that the column aligns even in textual form

| ----------------------------------------| -------------------------------------------------------------------- | -------------------------------------- |
| `ready_latency<int>`                    | non-negative integer                                                 | 0                                      |
| `bits_per_symbol<int>`                  | non-negative integer that divides the size of the data type          | 8                                      |
| `uses_valid<bool>`                      | boolean                                                              | `true`                                 |
| `first_symbol_in_high_order_bits<bool>` | boolean                                                              | `true`                                 |

> | `protocol`                              | `protocol_avalon_streaming` / `protocol_avalon_streaming_uses_ready` | `protocol_avalon_streaming_uses_ready` |

Reviewer: ```suggestion
| Property                                | Default Value                          |Valid Values
| ---                                     | ---                                    |---
| `ready_latency<int>`                    | 0                                      |non-negative integer
| `bits_per_symbol<int>`                  | 8                                      |non-negative integer that divides the size of the data type
| `uses_valid<bool>`                      | `true`                                 |boolean
| `first_symbol_in_high_order_bits<bool>` | `true`                                 |boolean
| `protocol`                              | `protocol_avalon_streaming_uses_ready` |`protocol_avalon_streaming` / `protocol_avalon_streaming_uses_ready`
```

> > **Note:** These properties may be specified in *any* order within the oneAPI `properties` object. Omitting a single property from the properties class instructs the compiler to assume the default value for that property (i.e., you can just define the properties you would like to change from the default). Omitting the properties template parameter entirely instructs the compiler to assume the default values for *all* properties.

Reviewer: ```suggestion
> **Note:** These properties may be specified in any order within the oneAPI `properties` list. Omitting a single property from the properties class instructs the compiler to assume the default value for that property, so you can just define the properties you would like to change from the default. Omitting the properties template parameter entirely instructs the compiler to assume the default values for *all* properties.
```

**technically** the `properties` thing is a *type*, but I don't think that's terribly important to drive home in this tutorial, so I recommend just say "property list"

Eric has given me feedback before that it's bad practice to use abbreviations like "i.e." and "e.g." in documentation.

#### Example 1.

> The following example declares a pipe to implement a streaming data interface with  `ready` and `valid` signals, using the defaults for all other parameters.

Reviewer: ```suggestion
The following example explicitly declares a pipe that implements a streaming data interface with  `ready` and `valid` signals, using the defaults for all other parameters.
```

```c++

> // Unique user-defined types

Reviewer: ```suggestion
// Forward declare the pipe name in the global scope. This is an FPGA best
// practice that reduces name mangling in the optimization reports.
```

class FirstPipeT;

// Pipe properties (listed here are the defaults; this achieves the same
// behavior as not specifying any of these properties)
using PipePropertiesT = decltype(sycl::ext::oneapi::experimental::properties(
    sycl::ext::intel::experimental::ready_latency<0>,
    sycl::ext::intel::experimental::bits_per_symbol<8>,
    sycl::ext::intel::experimental::uses_valid<true>,
    sycl::ext::intel::experimental::first_symbol_in_high_order_bits<true>,
    sycl::ext::intel::experimental::protocol_avalon_streaming_uses_ready));

using FirstPipeInstance = sycl::ext::intel::experimental::pipe<
    FirstPipeT,      // An identifier for the pipe
    int,             // The type of data in the pipe

>     8,               // The capacity of the pipe

Reviewer: ```suggestion
    8,               // Minimum capacity of the pipe (buffer depth)
```

    PipePropertiesT  // Customizable pipe properties
    >;
```

### Avalon Streaming Sideband Signals

You can enable Avalon streaming sideband signal support by using the special `StreamingBeat` struct provided by the `pipes_ext.hpp` header file (`sycl/ext/intel/prototype/pipes_ext.hpp`) as the data type to your pipe. Only the `StreamingBeat` struct generates sideband signals when used with a pipe.

> The `StreamingBeat` struct is templated on three parameters.

Reviewer: a table would be nice here, similar to above

- The first template parameter defines the type of the data being communicated through the interface.
- The second parameter is used to enable additional `startofpacket` (`sop`) and `endofpacket` (`eop`) signals to the Avalon interface.
- The third template parameter is used to enable the `empty` signal, which indicates the number of symbols that are empty during the `eop` cycle.

#### Example 2.

> The following example shows how to configure a `StreamingBeat` struct to use `sop`, `eop` and `empty` by setting the second and third template parameters to `true`. Using the `StreamingBeatT` type defined here in a pipe will result in additional `startofpacket`, `endofpacket`, and `empty` signals on the interface.

Reviewer: ```suggestion
The following example shows how to configure a `StreamingBeat` structure to use `sop`, `eop` and `empty` by setting the second and third template parameters to `true`. Using the `StreamingBeatT` type defined here in a pipe will result in `startofpacket`, `endofpacket`, and `empty` signals appearing on the resulting Avalon streaming interface.
```

```c++
#include <sycl/ext/intel/prototype/pipes_ext.hpp>

...

> using StreamingBeatDataT = sycl::ext::intel::experimental::StreamingBeat<unsigned char, true, true>;

Reviewer: ```suggestion
using StreamingBeatDataT = sycl::ext::intel::experimental::StreamingBeat<unsigned short, true, true>;
```

You said earlier not to write code like this :)

### On Linux*

1. Change to the sample directory.
2. Build the program for Intel® Agilex® 7 device family, which is the default.
   ```
   mkdir build
   cd build
   cmake ..
   ```
   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ```
   >
   > This tutorial is only intended for use in the IP Authoring flow and does not support targeting an explicit FPGA board variant and BSP.

3. Compile the design. (The provided targets match the recommended development flow.)

   1. Compile and run for emulation (fast compile time, targets emulated FPGA device).
      ```
      make fpga_emu
      ```
   2. Generate the optimization report.
      ```
      make report
      ```
   3. Compile and run for simulation (fast compile time, targets simulated FPGA device).
      ```
      make fpga_sim
      ```
   4. Compile for FPGA hardware (longer compile time, targets an FPGA device).
      ```

>       make fpga

Reviewer 2: The corresponding cmake target is missing

Author: fixed

## Run the `Streaming Data Interfaces` Tutorial

### On Linux

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   ./threshold_packets.fpga_emu
   ```
2. Run the sample on the FPGA simulator.
   ```
   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./threshold_packets.fpga_sim
   ```
	
### On Windows

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   threshold_packets.fpga_emu.exe
   ```
2. Run the sample on the FPGA simulator.
   ```
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1
   threshold_packets.fpga_sim.exe
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=
   ```

## Understanding the Tutorial

In `threshold_packets.cpp`, two pipes are declared for implementing the streaming input and streaming output interfaces on a kernel which thresholds pixel values in an image. The streams use start of packet and end of packet signals to determine the beginning and end of the image.

### Reading the Reports

> After compiling in the report flow, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information for these interfaces in the Details pane. This pane will identify that the read is coming from `InPixel`, and that the write is going to `OutPixel`, as well as verifying that both interfaces have a width of 32 bits (corresponding to the `StreamingBeatT` type) and depth of 8 (which is the capacity that each pipe was declared with).

Author: The report says the interface has a width of 32 bits (= `sizeof(StreamingBeatT)`), but the documentation says _"Using the StreamingBeat struct with the uses_packets template parameter set to true adds two additional 1-bit signals to the Avalon interface, start_of_packet (sop), and end_of_packet (eop)."_ Would that not mean the interface should be 16 + 2 = 18 bits wide? or does it do some packing?

With an 8-bit-wide data type the report shows a width of 24 bits.

Reviewer 2: Something to clarify with eng

Reviewer: that sounds like a report defect to me.
the report should show the interface width as 16 bits wide (since you declared `StreamingBeatT` to have `ac_int<16, false>` as the underlying type).

If an 8-bit-wide data type results in the report showing a 24-bit interface, it sounds like those 1-bit signals are being parsed inside the compiler as 8-bit signals, and those are trimmed down after the report is generated.

Author: Okay. I will open a case on this

> After compiling in the report flow, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information for these interfaces in the Details pane. This pane will identify that the read is coming from `InPixel`, and that the write is going to `OutPixel`, as well as verifying that both interfaces have a width of 32 bits (corresponding to the `StreamingBeatT` type) and depth of 8 (which is the capacity that each pipe was declared with).

Reviewer: ```suggestion
After compiling the `report` target, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information for these interfaces in the Details pane. This pane will identify that the read is coming from `InPixel`, and that the write is going to `OutPixel`, as well as verifying that both interfaces have a width of 32 bits (corresponding to the `StreamingBeatT` type) and depth of 8 (which is the capacity that each pipe was declared with).
```

# `Streaming Data Interfaces` Sample

This FPGA sample is a tutorial that demonstrates how to implement streaming data interfaces on an IP component. It is recommended that you review the [Component Interfaces Comparison](https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/C%2B%2BSYCL_FPGA/Tutorials/Features/component_interfaces_comparison) tutorial before continuing with this one.

| Area                  | Description
|:--                    |:--
| What you will learn   | How to use pipes to implement streaming data interfaces on an IP component
| Time to complete      | 30 minutes
| Category              | Concepts and Functionality

## Purpose

Pipes are a first-in first-out (FIFO) buffer construct that provide data links between elements of a design. They are accessed through read and write APIs without the notion of a memory address or pointers to elements within the FIFO.

> The concept of a pipe is an intuitive mechanism for specifying streaming data interfaces on an IP component. This tutorial demonstrates how to declare and configure a pipe to configure an Avalon streaming interface.

Reviewer: ```suggestion
The concept of a pipe is an intuitive mechanism for specifying streaming data interfaces on an IP component. This tutorial demonstrates how to use the pipe API to configure an Avalon streaming interface.
```

   tier3("Tier 3: Explore the Advanced Techniques")
   tier4("Tier 4: Explore the Reference Designs")

   tier1 --> tier2 --> tier3 --> tier4

   style tier1 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier2 fill:#f96,stroke:#333,stroke-width:1px,color:#fff
   style tier3 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier4 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
```

Find more information about how to navigate this part of the code samples in the [FPGA top-level README.md](/DirectProgramming/C++SYCL_FPGA/README.md).
You can also find more information about [troubleshooting build errors](/DirectProgramming/C++SYCL_FPGA/README.md#troubleshooting), [running the sample on the Intel® DevCloud](/DirectProgramming/C++SYCL_FPGA/README.md#build-and-run-the-samples-on-intel-devcloud-optional), [using Visual Studio Code with the code samples](/DirectProgramming/C++SYCL_FPGA/README.md#use-visual-studio-code-vs-code-optional), [links to selected documentation](/DirectProgramming/C++SYCL_FPGA/README.md#documentation), and more.

## Key Implementation Details

### Configuring a Pipe to Implement a Streaming Data Interface

> Each individual pipe is a function scope class declaration of the templated `pipe` class. It takes two mandatory and two optional parameters, as summarized in Table 1.

Reviewer: ```suggestion
Each pipe is a class declaration of the templated `pipe` class. A pipe declaration takes two mandatory and two optional parameters, as summarized in Table 1.
```

#### Table 1. Template Parameters of the `pipe` Class

| Template Parameter                      | Description
| ---                                     | ---

> | `name`                                  | A user-defined type that differentiates this particular pipe from the others and provides a name for the interface in the RTL.

Reviewer: ```suggestion
| `name`                                  | A user-defined type that uniquely identifies this pipe. The name of this type is also used to identify the interface in the generated RTL.
```

| `dataT`                                 | The datatype of elements carried by the interface.*

> | `min_capacity`                          | A non-negative integer representing the capacity of the buffer on the input. This can help avoid some amount of bubbles in the pipeline in case the component itself stalls. This parameter is optional, and defaults to 0.

Reviewer: ```suggestion
| `dataT`                                 | The type of data that passes through the pipe*. This is the data type that is read during a successful pipe read() operation, or written during a successful pipe write() operation. The type must have a standard layout and be trivially copyable.
| `min_capacity`                          | User-defined minimum number of words (in units of `dataT`) that the pipe must be able to store without any being read out. This parameter is optional, and defaults to 0.
```

https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/2023-2/the-pipe-class-and-its-use.html

| `properties`                            | An unordered list of SYCL properties that define additional semantic properties for a pipe. This parameter is optional.**

> :warning: * There is currently a known issue with using structure types whose first field is 8-bits wide, and hence also data types which are themselves 8-bits wide (for example, `unsigned char`). For the time being, please use a wider datatype where applicable (for example, `unsigned short`).

> Below is a summary of all relevant SYCL properties which can be applied to a `pipe` using the fourth template parameter. Please note that this table is not complete; see the [FPGA Optimization Guide for Intel® oneAPI Toolkits](https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/current/host-pipe-declaration.html) for more information on how to use pipes in other applications).

Reviewer: ```suggestion
Below is a summary of all relevant SYCL properties which can be applied to a `pipe` using the `properties` template parameter. Please note that this table is not complete; see the [FPGA Optimization Guide for Intel® oneAPI Toolkits](https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/current/host-pipe-declaration.html) for more information on how to use pipes in other applications.
```

it might be nice to link to the avalon interface specifications https://www.intel.com/content/www/us/en/docs/programmable/683091/22-3/st-interface-properties.html but maybe that will make this too long and that should really be in the manual.

Author: I just added a link to the Avalon interface spec after the table:

_See [this page](https://www.intel.com/content/www/us/en/docs/programmable/683091/current/st-interface-properties.html) for more information on the Avalon interface specifications._

Let me know if you think we should keep it.

#### Table 2. Properties used to Configure a Pipe to Implement a Streaming Data Interface

| Property                                | Default Value                          | Valid Values
| ---                                     | ---                                    | ---
| `ready_latency<int>`                    | 0                                      | non-negative integer
| `bits_per_symbol<int>`                  | 8                                      | non-negative integer that divides the size of the data type
| `uses_valid<bool>`                      | `true`                                 | boolean
| `first_symbol_in_high_order_bits<bool>` | `true`                                 | boolean
| `protocol`                              | `protocol_avalon_streaming_uses_ready` |`protocol_avalon_streaming` / `protocol_avalon_streaming_uses_ready`

> ** Omitting a single property from the properties class instructs the compiler to assume the default value for that property, so you can just define the properties you would like to change from the default. Omitting the properties template parameter entirely instructs the compiler to assume the default values for all properties.

Reviewer: ```suggestion
> *Note*: ** Omitting a single property from the properties class instructs the compiler to assume the default value for that property, so you can just define the properties you would like to change from the default. Omitting the properties template parameter entirely instructs the compiler to assume the default values for all properties.
```
This can be moved right under the note about the unsigned char bug

#### Example 1.

The following example explicitly declares a pipe that implements a streaming data interface with  `ready` and `valid` signals, using the defaults for all other parameters.

```c++
// Forward declare the pipe name in the global scope. This is an FPGA best
// practice that reduces name mangling in the optimization reports.
class FirstPipeT;

// Pipe properties (listed here are the defaults; this achieves the same
// behavior as not specifying any of these properties)
using PipePropertiesT = decltype(sycl::ext::oneapi::experimental::properties(
    sycl::ext::intel::experimental::ready_latency<0>,
    sycl::ext::intel::experimental::bits_per_symbol<8>,
    sycl::ext::intel::experimental::uses_valid<true>,
    sycl::ext::intel::experimental::first_symbol_in_high_order_bits<true>,

>     sycl::ext::intel::experimental::protocol_avalon_streaming_uses_ready));

Reviewer: Is it `decltype (properties { ready_latency...})`? I've seen the `properties` type configured with curly braces and regular parentheses.

Author: interesting, both seem to work... however our documentation shows it with regular parentheses, so I will probably stick to that. (e.g. https://www.intel.com/content/www/us/en/docs/oneapi-fpga-add-on/optimization-guide/2023-2/host-pipe-declaration.html)

using FirstPipeInstance = sycl::ext::intel::experimental::pipe<
    FirstPipeT,      // An identifier for the pipe
    int,             // The type of data in the pipe
    8,               // Minimum capacity of the pipe (buffer depth)
    PipePropertiesT  // Customizable pipe properties

>     >;

Reviewer: ```suggestion
using FirstPipe = sycl::ext::intel::experimental::pipe<
    FirstPipeT,      // An identifier for the pipe
    int,             // The type of data in the pipe
    8,               // Minimum capacity of the pipe (buffer depth)
    PipePropertiesT  // Customizable pipe properties
    >;
```

it's not technically an instance

```

### Avalon Streaming Sideband Signals

> You can enable Avalon streaming sideband signal support by using the special `StreamingBeat` structure provided by the `pipes_ext.hpp` header file (`sycl/ext/intel/prototype/pipes_ext.hpp`) as the data type to your pipe. Only the `StreamingBeat` structure generates sideband signals when used with a pipe.

Reviewer: ```suggestion
Pipes support a subset of Avalon streaming sideband signals. You can add these to your pipe interface by using the special `StreamingBeat` structure provided by the `pipes_ext.hpp` header file (`sycl/ext/intel/prototype/pipes_ext.hpp`) as the `dataT` of your pipe. Only the `StreamingBeat` structure generates sideband signals when used with a pipe.
```

The `StreamingBeat` structure is templated on three parameters, as summarized in Table 3.

#### Table 3. Template Parameters of the `StreamingBeat` Structure

| Template Parameter                      | Description
| ---                                     | ---

> | `dataT`                                 | The datatype of elements carried by the interface.

Reviewer: ```suggestion
| `dataT`                                 | The datatype of elements carried by the `data` signal of the Avalon streaming interface.
```

> | `uses_packets`                          | A boolean to indicate whether to enable additional `startofpacket` (`sop`) and `endofpacket` (`eop`) signals to the Avalon interface.

Reviewer: ```suggestion
| `uses_packets`                          | A boolean that indicates whether to enable the `startofpacket` (`sop`) and `endofpacket` (`eop`) sideband signals on the Avalon streaming interface.
```

> | `uses_empty`                            | A boolean to indicate whether to enables the `empty` signal, which indicates the number of symbols that are empty during the `eop` cycle.

Reviewer: ```suggestion
| `uses_empty`                            | A boolean that indicates whether to enable the `empty` sideband signal on the Avalon streaming interface
```

      ```
   3. Compile and run for simulation (fast compile time, targets simulated FPGA device).
      ```
      nmake fpga_sim
      ```
   4. Compile for FPGA hardware (longer compile time, targets an FPGA device).
      ```
      nmake fpga
      ```
> **Note**: If you encounter any issues with long paths when compiling under Windows*, you may have to create your ‘build’ directory in a shorter path, for example c:\samples\build.  You can then run cmake from that directory, and provide cmake with the full path to your sample directory.

## Run the `Streaming Data Interfaces` Tutorial

### On Linux

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   ./threshold_packets.fpga_emu
   ```
2. Run the sample on the FPGA simulator.
   ```
   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./threshold_packets.fpga_sim
   ```
	
### On Windows

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   threshold_packets.fpga_emu.exe
   ```
2. Run the sample on the FPGA simulator.
   ```
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1
   threshold_packets.fpga_sim.exe
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=
   ```

## Understanding the Tutorial

> In `threshold_packets.cpp`, two pipes are declared for implementing the streaming input and streaming output interfaces on a kernel which thresholds pixel values in an image. The streams use start of packet and end of packet signals to determine the beginning and end of the image.

Reviewer: ```suggestion
In `threshold_packets.cpp`, two pipes are declared to implement the input and output streaming interfaces on a kernel which thresholds pixel values in an image. The streams use `startofpacket` and `endofpacket` signals to determine the beginning and end of the image.
```

### Reading the Reports

> After compiling the `report` target, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information for these interfaces in the Details pane. This pane will identify that the read is coming from `InPixel`, and that the write is going to `OutPixel`, as well as verifying that both interfaces have a width of 32 bits (corresponding to the `StreamingBeatT` type) and depth of 8 (which is the capacity that each pipe was declared with).

Reviewer: ```suggestion
After compiling the `report` target, locate and open the `report.html` file in the `threshold_packets.report.prj/reports/` directory. Under the `Threshold` kernel in the System Viewer, the streaming in and streaming out interfaces can be seen, shown by the pipe read and pipe write nodes respectively. Clicking on either of these nodes gives further information about these interfaces in the 'details' pane. The 'details' pane will identify that the read is coming from `InPixel`, and that the write is going to `OutPixel`, as well as verifying that both interfaces have a width of 32 bits (corresponding to size of the `StreamingBeatT` type) and depth of 8 (which is the capacity that each pipe was declared with).
```
